keps/sig-node/20190226-pod-overhead.md

```diff
@@ -0,0 +1,300 @@
+approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Sandbox runtimes introduce a non-negligible overhead at the pod level which must be accounted for
+effective scheduling, resource quota management, and constraining.
+
+## Motivation
+
+Pods have some resource overhead. In our traditional linux container (Docker) approach,
+the accounted overhead is limited to the infra (pause) container, but also invokes some
+overhead accounted to various system components including: Kubelet (control loops), Docker,
+kernel (various resources), fluentd (logs). The current approach is to reserve a chunk
+of resources for the system components (system-reserved, kube-reserved, fluentd resource
+request padding), and ignore the (relatively small) overhead from the pause container, but
+this approach is heuristic at best and doesn't scale well.
+
+With sandbox pods, the pod overhead potentially becomes much larger, maybe O(100mb). For
+example, Kata containers must run a guest kernel, kata agent, init system, etc. Since this
+overhead is too big to ignore, we need a way to account for it, starting from quota enforcement
+and scheduling.
+
+### Goals
+
+* Provide a mechanism for accounting pod overheads which are specific to a given runtime solution
+
+### Non-Goals
+
+* Making runtimeClass selections
+
+## Proposal
+
+Augment the RuntimeClass custom resource definition and the `PodSpec` to introduce the field
+`Overhead *ResourceRequirements`. This field represents the overhead associated with running a pod
+for a given runtimeClass.  A mutating admission controller is introduced which will update the `Overhead`
+field in the workload's `PodSpec` to match what is provided for the selected RuntimeClass, if one is specified.
+
+Kubelet's creation of the pod cgroup will be calculated as the sum of container `ResourceRequirements` fields,
+plus the Overhead associated with the pod.
+
+The scheduler, resource quota handling and Kubelet's pod cgroup creation will take Overhead into account, as well
+as the sum of the pod's container requests.
+
+### API Design
+
+#### Pod overhead
+Introduce a Pod.Spec.Resources field on the pod to specify the pods overhead.
+
+```
+Pod {
+  Spec PodSpec {
+    // Overhead is the resource overhead consumed by the Pod, not including
+    // container resource usage. Users should leave this field unset.
```

Comment:
Specs must not include "should".   What happens if a user specifies this as way too big?  What happens if the specify this as way too small?  Is this a vector for DoS or other attacks on components which trust this?

As an alternative, could wehave Kubelet write this to status (e.g. read from runtime class, write to status) ?

Reply:
if it remains in spec (which if we do this, i think it should), ability to write the field value should be controlled via a subresource similar to binding.  we would need an identity that is allowed to do just the overhead field setting similar to how schedulers can do bind.

Reply:
Thanks @thockin @derekwaynecarr  -- I'll need to take a look at binding to better understand.

Conceptually, any provided value should be mutated at admission time by a runtimeClass controller utilizing on of the following:
1) original value ignored and overwritten with what is specified for the given runtimeClass, or
2) The larger of initial value or runtimeClass value should be written.

Reply:
Other options include:

3. Only mutate the PodOverhead if it is unset (defaulting)
4. If an inconsistent value is already set, reject the pod

I'm leaning towards option 4. The advantage of 4 is we don't think people should be setting this value anyway, so just fail early. If we go with (1) there's a risk of it becoming confusing that the pod is running with a different overhead than what was originally set.

Reply:
@tallclair Thanks.

For option 4, can you expand on the definition of an inconsistent value?

Reply:
updated  the document to propose rejection of the pod if an inconsistent value is provided.  Failing early makes most sense.